core/src/main/java/com/amazon/opendistroforelasticsearch/sql/planner/physical/DedupeOperator.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  public DedupeOperator(
+      PhysicalPlan input,
+      List<Expression> dedupeList,
+      Integer allowedDuplication,
+      Boolean keepEmpty,
+      Boolean consecutive) {
+    this.input = input;
+    this.dedupeList = dedupeList;
+    this.allowedDuplication = allowedDuplication;
+    this.keepEmpty = keepEmpty;
+    this.consecutive = consecutive;
+    this.deduper = this.consecutive ? new ConsecutiveDeduper<>() : new HistoricalDeduper<>();
+  }
+
+  @Override
+  public <R, C> R accept(PhysicalPlanNodeVisitor<R, C> visitor, C context) {
+    return visitor.visitDedupe(this, context);
+  }
+
+  @Override
+  public List<PhysicalPlan> getChild() {
+    return Collections.singletonList(input);
+  }
+
+  @Override
+  public boolean hasNext() {
+    while (input.hasNext()) {
+      ExprValue next = input.next();
+      if (keep(next)) {
+        this.next = next;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  @Override
+  public ExprValue next() {
+    return this.next;
+  }
+
+  /**
+   * Test the {@link ExprValue} should be keep or ignore
+   *
+   * <p>If any value evaluted by {@link DedupeOperator#dedupeList} is NULL or MISSING, then the *
+   * return value is decided by keepEmpty option, default value is ignore.
+   *
+   * @param value
+   * @return true: keep, false: ignore
+   */
+  public boolean keep(ExprValue value) {
+    BindingTuple bindingTuple = value.bindingTuples();
+    ImmutableList.Builder<ExprValue> dedupeKeyBuilder = new ImmutableList.Builder<>();
+    for (Expression expression : dedupeList) {
+      ExprValue exprValue = expression.valueOf(bindingTuple);
+      if (NULL_OR_MISSING.test(exprValue)) {
+        return keepEmpty;
+      }
+      dedupeKeyBuilder.add(exprValue);
+    }
+    List<ExprValue> dedupeKey = dedupeKeyBuilder.build();
+    int seenTimes = deduper.seenTimes(dedupeKey);
+    return seenTimes <= allowedDuplication;
+  }
+
+  /**
+   * Return how many times the dedupeKey has been seen before. The side effect is the seen times
+   * will add 1 times after calling this function.
+   *
+   * @param <K> dedupe key
+   */
+  interface Deduper<K> {
+
+    int seenTimes(K dedupeKey);
+  }
+
+  /** The Historical Deduper monitor the duplicated element with all the seen value. */
+  static class HistoricalDeduper<K> implements Deduper<K> {
+    private final Map<K, Integer> seenMap = new ConcurrentHashMap<>();
+
+    @Override
+    public int seenTimes(K dedupeKey) {
+      seenMap.putIfAbsent(dedupeKey, 0);
+      return seenMap.computeIfPresent(dedupeKey, (k, v) -> v + 1);
+    }
+  }
+
+  /**
+   * The Consecutive Deduper monitor the duplicated element with consecutive seen value. It means
+   * only the consecutive duplicated value will be counted.
+   */
+  static class ConsecutiveDeduper<K> implements Deduper<K> {
+    private K lastSeenDedupeKey = null;
+    private Integer consecutiveCount = 0;
+
+    @Override
+    public int seenTimes(K dedupeKey) {
```

Comment:
Notice that `HistoricalDeduper` is using `ConcurrentHashMap`, so just wonder if thread safe is also needed here by synchronized or AtomicReference?

Reply:
Done. Refactor the Deduper. Get rid of hierarchy

Reply:
Thanks!